Node locks up after partitions

While trying to fix a bug in the tests for #2152, I managed to lock up dgraph into a state where every request to one node (n1) timed out, despite the process running on all nodes. Here's the complete logs and data file from all five nodes.

[dgraph-n1-lockup.zip](https://github.com/dgraph-io/dgraph/files/1854491/dgraph-n1-lockup.zip)

This occurs on

Dgraph version   : v1.0.4
Commit SHA-1     : 807976c
Commit timestamp : 2018-03-22 14:55:24 +1100
Branch           : master

and involves a series of network partitions; it looks as if the final partition healing might have left n1 in a state where it believed the leader was... possibly a node which was not the leader?

```
2018/03/27 12:22:51 node.go:344: Error while sending message to node with addr: n2:7080, err: rpc error: code = DeadlineExceeded desc = context deadline exceeded
2018/03/27 12:22:51 groups.go:702: Error in oracle delta stream. Error: rpc error: code = Unknown desc = Node is no longer leader.
2018/03/27 12:22:51 Error while retrieving timestamps: rpc error: code = Unknown desc = Assigning IDs is only allowed on leader.. Will retry...
```

To reproduce this, run with Jepsen e31b29d1a5302766c2c83454eeed9124ef9820f5:

```
lein run test --package-url https://transfer.sh/TjHBo/dgraph-linux-amd64.tar.gz --force-download -w set --time-limit 300 --concurrency 2n --nemesis partition-random-halves
```

This appears to be a semi-rare fault; I've only seen it once so far.

--- a/dgraph/cmd/zero/raft.go
+++ b/dgraph/cmd/zero/raft.go
@@ -404,8 +404,7 @@ func (n *node) applyConfChange(e raftpb.Entry) {
 
 func (n *node) triggerLeaderChange() {
 	n.server.triggerLeaderChange()
-	m := &intern.Member{Id: n.Id, Addr: n.RaftContext.Addr, Leader: n.AmLeader()}
-	go n.proposeAndWait(context.Background(), &intern.ZeroProposal{Member: m})
+	n.server.updateZeroLeader()
 }
 
 func (n *node) initAndStartNode(wal *raftwal.Wal) error {

--- a/dgraph/cmd/zero/raft.go
+++ b/dgraph/cmd/zero/raft.go
@@ -404,6 +404,8 @@ func (n *node) applyConfChange(e raftpb.Entry) {
 
 func (n *node) triggerLeaderChange() {
 	n.server.triggerLeaderChange()
+	// We update leader information on each node without proposal. This
+	// function is called on all nodes on leader change.
 	n.server.updateZeroLeader()
 }
 

--- a/dgraph/cmd/zero/zero.go
+++ b/dgraph/cmd/zero/zero.go
@@ -203,6 +203,19 @@ func (s *Server) storeZero(m *intern.Member) {
 	s.state.Zeros[m.Id] = m
 }
 
+func (s *Server) updateZeroLeader() {
+	s.Lock()
+	defer s.Unlock()
+	leader := s.Node.Raft().Status().Lead
+	for _, m := range s.state.Zeros {
+		if m.Id == leader {
+			m.Leader = true
+		} else {
+			m.Leader = false
+		}
+	}
+}
+
 func (s *Server) removeZero(nodeId uint64) {
 	s.Lock()
 	defer s.Unlock()

--- a/worker/groups.go
+++ b/worker/groups.go
@@ -247,6 +247,9 @@ func (g *groupi) applyState(state *intern.MembershipState) {
 	x.AssertTrue(state != nil)
 	g.Lock()
 	defer g.Unlock()
+	// We don't update state if we get any old state. Counter stores the raftindex of
+	// last update. For leader changes at zero since we don't propose, state can get
+	// updated at same counter value. So ignore only if counter is less.
 	if g.state != nil && g.state.Counter > state.Counter {
 		return
 	}

--- a/worker/groups.go
+++ b/worker/groups.go
@@ -247,7 +247,7 @@ func (g *groupi) applyState(state *intern.MembershipState) {
 	x.AssertTrue(state != nil)
 	g.Lock()
 	defer g.Unlock()
-	if g.state != nil && g.state.Counter >= state.Counter {
+	if g.state != nil && g.state.Counter > state.Counter {
 		return
 	}
 

--- a/worker/predicate_move.go
+++ b/worker/predicate_move.go
@@ -134,7 +134,10 @@ func movePredicateHelper(ctx context.Context, predicate string, gid uint32) erro
 		}
 		prevKey = prevKey[:len(key)]
 		copy(prevKey, key)
-		l, err := posting.ReadPostingList(prevKey, it)
+
+		nkey := make([]byte, len(key))
+		copy(nkey, key)
+		l, err := posting.ReadPostingList(nkey, it)
 		if err != nil {
 			return err
 		}

